csvy/readers.py

```diff
@@ -270,3 +271,52 @@ def read_to_list(
             data.append(row)
 
     return data, header
+
+
+def read_into_dict(file_path: Path | str):
+    """Read a CSVY file and return its contents as a dictionary and metadata.
+
+    This function assumes that the first row of the data section contains the
+    column names.
+
+    Args:
+        file_path (str): The path to the CSVY file to read.
+
+    Returns:
+        tuple: A tuple containing:
+            - data_dict (dict): A dictionary where keys are column names and
+              values are lists of entries.
+            - metadata (dict): A dictionary containing the extracted metadata
+              from the CSVY file.
+    """
+    with open(file_path) as f:
+        lines = f.readlines()
+
+    metadata = {}
+    data = []
+    section = None
+
+    for line in lines:
+        line = line.strip()
+        if line.startswith("# ---"):
+            if section is None:
+                section = "metadata"
+            elif section == "metadata":
+                section = "data"
+            else:
+                break
+            continue
+
+        if section == "metadata":
+            if ":" in line:
+                key, value = line.split(":", 1)
+                metadata[key.strip()] = value.strip()
+        elif section == "data":
+            if line:
+                data.append(line.split(","))
+
+    if data:
+        data = list(map(list, zip_longest(*data, fillvalue="")))
+        data_dict = {data[i][0]: data[i][1:] for i in range(len(data))}
+        return data_dict, metadata
```

Comment:
I believe this section of code should work, but I haven't checked it yet because of the other issues.